File: java/LetterCombinationsOfAPhoneNumber.java
import java.util.ArrayList;

public class LetterCombinationsOfAPhoneNumber {
    private StringBuilder strBuilder;
    private ArrayList<String> ansList;
    private String[] codes = new String[] {"", "", "abc", "def", "ghi", 
        "jkl", "mno", "pqrs", "tuv", "wxyz"};

    private void searchCombinations(String digits, int ptr) {
        if (ptr >= digits.length()) {
            ansList.add(strBuilder.toString());
            return;
        }
        int num = digits.charAt(ptr) - '0';
        int n = codes[num].length();
        for (int i = 0; i < n; ++i) {
            strBuilder.append(codes[num].charAt(i));
            searchCombinations(digits, ptr + 1);
            strBuilder.deleteCharAt(strBuilder.length() - 1);
        }
    }

    public ArrayList<String> letterCombinations(String digits) {
        strBuilder = new StringBuilder();
        ansList = new ArrayList<String>();
        if (digits == null) {
            return ansList;
        }
        searchCombinations(digits, 0);
        return ansList;
    }
}

File: cc/letter_combinations_of_a_phone_number.cc
#include "base_include.h"

class LetterCombinationsOfAPhoneNumber {
public:
    vector<string> letterCombinations(const string &digits) {
        int n = digits.size();
        vector<string> res;
        if (0 == n) {
            res.push_back("");
            return res;
        }
        int d = digits[0] - '0';
        int m = strs[d].size();
        vector<string> sub = letterCombinations(digits.substr(1));
        for (int i = 0; i < m; ++i) {
            for (auto &s : sub) {
                res.push_back(strs[d][i] + s);
            }
        }
        return res;
    }
private:
    string strs[10]{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
};

int main(int argc, char *argv[]) {
    return 0;
}
